camus-api/src/main/java/com/linkedin/camus/schemaregistry/CachedSchemaRegistry.java

```diff
@@ -2,86 +2,153 @@
+  private static final String GETSCHEMA_BYID_MIN_INTERVAL_SECONDS = "getschema.byid.min.interval.seconds";
+  private static final String GETSCHEMA_BYID_MIN_INTERVAL_SECONDS_DEFAULT = "0";
+
+  private final SchemaRegistry<S> registry;
+  private final ConcurrentHashMap<CachedSchemaTuple, S> cachedById;
+  private final ConcurrentHashMap<String, SchemaDetails<S>> cachedLatest;
+  private int getByIdMaxRetries;
+  private int getByIdMinIntervalSeconds;
+  private ConcurrentMap<String, FailedFetchHistory> failedFetchHistories;
+
+  private class FailedFetchHistory {
+    private int numOfAttempts;
+    private long previousAttemptTime;
+
+    private FailedFetchHistory(int numOfnumOfAttempts, long previousAttemptTime) {
+      this.numOfAttempts = numOfnumOfAttempts;
+      this.previousAttemptTime = previousAttemptTime;
+    }
+  }
+
+  public void init(Properties props) {
+    this.getByIdMaxRetries = Integer.parseInt(props.getProperty(
+        GETSCHEMA_BYID_MAX_RETIRES, GETSCHEMA_BYID_MAX_RETIRES_DEFAULT));
+    this.getByIdMinIntervalSeconds = Integer.parseInt(props.getProperty(
+        GETSCHEMA_BYID_MIN_INTERVAL_SECONDS,
+        GETSCHEMA_BYID_MIN_INTERVAL_SECONDS_DEFAULT));
+  }
+
+  public CachedSchemaRegistry(SchemaRegistry<S> registry) {
+    this.registry = registry;
+    this.cachedById = new ConcurrentHashMap<CachedSchemaTuple, S>();
+    this.cachedLatest = new ConcurrentHashMap<String, SchemaDetails<S>>();
+    this.getByIdMaxRetries = Integer
+        .parseInt(GETSCHEMA_BYID_MAX_RETIRES_DEFAULT);
+    getByIdMinIntervalSeconds = Integer
+        .parseInt(GETSCHEMA_BYID_MIN_INTERVAL_SECONDS_DEFAULT);
+    this.failedFetchHistories = new ConcurrentHashMap<String, FailedFetchHistory>();
+  }
+
+  public String register(String topic, S schema) {
+    return registry.register(topic, schema);
+  }
+
+  public S getSchemaByID(String topic, String id) {
+    CachedSchemaTuple cacheKey = new CachedSchemaTuple(topic, id);
+    S schema = cachedById.get(cacheKey);
+    if (schema == null) {
+      if (shouldFetchFromSchemaRegistry(id)) {
+        schema = fetchFromSchemaRegistry(topic, id);
+      } else {
+        throw new SchemaNotFoundException();
+      }
+    }
+    return schema;
+  }
+
+  private boolean shouldFetchFromSchemaRegistry(String id) {
+    if (!failedFetchHistories.containsKey(id)) {
+      return true;
+    }
+    FailedFetchHistory failedFetchHistory = failedFetchHistories.get(id);
+    boolean maxRetriesNotExceeded = failedFetchHistory.numOfAttempts < getByIdMaxRetries;
+    boolean minRetryIntervalSatisfied = System.nanoTime()
+        - failedFetchHistory.previousAttemptTime >= TimeUnit.SECONDS
+        .toNanos(getByIdMinIntervalSeconds);
+    return maxRetriesNotExceeded && minRetryIntervalSatisfied;
+  }
+
+  private S fetchFromSchemaRegistry(String topic, String id) {
+    try {
+      S schema = registry.getSchemaByID(topic, id);
+      return schema;
+    } catch (SchemaNotFoundException e) {
```

Comment:
I did a global search and there's no match for 'SchemaRegistryClientException' anywhere (no definition or usage). Also, the SchemaRegistry interface Javadoc says it throws SchemaNotFoundException.